File: 03_utqiagvik/11_installation_maps.R
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Utqiagvik map results
# Author: Timm Nawrocki, Alaska Center for Conservation Science
# Last Updated: 2026-04-24
# Usage: Must be executed in a R 4.4.3+ installation.
# Description: 'Utqiagvik map results' creates a map figure for publication that shows the imagery, 0.5 m vegetation types, and 2 m vegetation types.
# ---------------------------------------------------------------------------

# Import required libraries
library(dplyr)
library(fs)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(ggspatial)
library(sf)
library(terra)
library(tidyterra)

# Set global font size
font_size = 19

#### SET UP DIRECTORIES, FILES, AND FIELDS
####____________________________________________________

# Set root directory
drive = 'C:'
root_folder = 'ACCS_Work'

# Define folder structure
project_folder = path(drive, root_folder, 'Projects/VegetationEcology/DoD_Navy_Arctic/Data')
imagery_folder = path(project_folder, 'Data_Input/imagery_data')
region_folder = path(project_folder, 'Data_Input/region_data')
vegetation_folder = path(project_folder, 'Data_Output/vegetation_data')
output_folder = path(project_folder, 'Data_Output/summary_results')

# Define input files
imagery_input = path(imagery_folder, 'Utqiagvik_Imagery_20240710_0.5m_3338.tif')
installation_input = path(region_folder, 'Barrow_Installation_Main_3338.shp')
coast_input = path(region_folder, 'Utqiagvik_CoastalZone_3338.shp')
vegtype_input = path(vegetation_folder, 'Utqiagvik_Vegetation_2024_0.5m_3338.tif')
vegcomplex_input = path(vegetation_folder, 'Utqiagvik_Vegetation_2024_900mmu_2m_3338.tif')

# Define output files
figure_output = path(output_folder, 'Utqiagvik_VegetationMaps_2024.jpg')

#### DEFINE FUNCTIONS
####____________________________________________________

# Define a function to append the colormap color values to the raster attribute table labels
get_color_palette = function(input_raster, color_path) {
  # Read raw lines from the colormap file
  raw_lines = readLines(color_path)
  
  # Keep all lines that start with a digit
  color_lines = raw_lines[grepl('^[0-9]', raw_lines)]
  
  # Add R, G, B values from the colormap to a table
  rgb_data = read.table(text = color_lines, header = FALSE)
  
  # Convert RGB columns to hex codes
  hex_colors = rgb(rgb_data[,2], rgb_data[,3], rgb_data[,4], maxColorValue = 255)
  
  # Create a data frame mapping VALUE to Hex
  color_data = data.frame(VALUE = rgb_data[,1], hex = hex_colors)
  
  # Extract the Raster Attribute Table
  attribute_data = levels(input_raster)[[1]]
  
  # Merge colors to the attribute data by VALUE
  merged_data = merge(attribute_data, color_data, by = 'VALUE')
  
  # Create named vector mapping LABEL to hex color for ggplot
  palette = setNames(merged_data$hex, merged_data$LABEL)
  return(palette)
}

#### PREPARE VECTOR DATA & DEFINE MAP EXTENT
####____________________________________________________

# Read shapefiles and enforce EPSG:3338
coast_data = st_read(coast_input) %>% 
  st_transform(crs = 3338)
installation_data = st_read(installation_input) %>% 
  st_transform(crs = 3338)

# Buffer the installation polygon by 50 meters
installation_buffered = st_buffer(installation_data, dist = 50)

# Create a bounding box based on the 50 m buffered installation bounds
bounding_box = st_bbox(installation_buffered)
bounding_3338 = st_as_sfc(bounding_box)

# Extract x and y limits from the buffered bounding box
x_limits = c(bounding_box['xmin'], bounding_box['xmax'])
y_limits = c(bounding_box['ymin'], bounding_box['ymax'])

#### PREPARE RASTER DATA & COLORMAPS
####____________________________________________________

# Read and crop imagery
imagery_raster = rast(imagery_input)
imagery_crop = crop(imagery_raster, bounding_3338)

# Read and crop vegetation type map
vegtype_raster = rast(vegtype_input)
vegtype_crop = crop(vegtype_raster, bounding_3338)

# Read and crop vegetation type-complex map
vegcomplex_raster = rast(vegcomplex_input)
vegcomplex_crop = crop(vegcomplex_raster, bounding_3338)

# Generate palettes from .clr files
type_palette = get_color_palette(vegtype_crop, paste0(vegtype_input, '.clr'))
complex_palette = get_color_palette(vegcomplex_crop, paste0(vegcomplex_input, '.clr'))

# Combine the two color palettes and remove duplicate types
color_palette = c(type_palette, complex_palette)
color_palette = color_palette[!duplicated(names(color_palette))]

# Extract the labels present in the cropped raster extents
type_present = na.omit(unique(values(vegtype_crop, dataframe = TRUE))[[1]])
complex_present = na.omit(unique(values(vegcomplex_crop, dataframe = TRUE))[[1]])

# Combine the unique labels from both cropped rasters and sort alphabetically
type_labels = sort(unique(c(type_present, complex_present)))

#### BUILD INDIVIDUAL MAP PANELS
####____________________________________________________

# Dynamically find the band index numbers for a CIR false color visualization
r_band = match('b04_nir', names(imagery_crop))
g_band = match('b03_red', names(imagery_crop))
b_band = match('b02_green', names(imagery_crop))

# Create imagery map panel (bottom left)
imagery_plot = ggplot() +
  # Add imagery raster as CIR false color
  geom_spatraster_rgb(
    data = imagery_crop, 
    r = r_band, 
    g = g_band, 
    b = b_band, 
    max_col_value = 220,
    stretch = 'lin'
  ) +
  # Add coastal zone overlay
  geom_sf(data = coast_data, color = 'black', fill = NA, linewidth = 0.8) +
  # Add installation boundary overlay
  geom_sf(data = installation_data, color = 'yellow', fill = NA, linewidth = 0.8) +
  # Adjust the plot to the horizontal and vertical installation bounds
  coord_sf(
    crs = st_crs(3338),
    xlim = x_limits,
    ylim = y_limits,
    expand = FALSE
  ) +
  # Add plot title
  ggtitle('c. High-resolution (0.5 m) CIR imagery') +
  # Add label with semi-transparent background
  annotate('label', 
           x = x_limits[1] + 20,
           y = y_limits[1] + 20, 
           label = 'Imagery © 2020 Maxar Technologies Inc.', 
           hjust = 0, vjust = 0, 
           size = 6, 
           color = 'black', 
           fill = alpha('white', 0.7),
           label.size = 0) +
  # Add plot styling
  theme_minimal() +
  theme(plot.margin = margin(2,2,2,2),
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_line(colour = 'gray'),
        plot.title = element_text(size = font_size)
  )

# Create vegetation type map panel (top left)
vegtype_plot = ggplot() +
  # Add vegetation type raster
  geom_spatraster(data = vegtype_crop) +
  scale_fill_manual(values = color_palette,               
                    breaks = type_labels,           
                    na.translate = FALSE, 
                    name = 'Vegetation Class') +
  # Add coastal zone overlay
  geom_sf(data = coast_data, color = 'black', fill = NA, linewidth = 0.8) +
  # Add installation boundary overlay
  geom_sf(data = installation_data, color = 'yellow', fill = NA, linewidth = 0.8) +
  # Adjust the plot to the horizontal and vertical installation bounds
  coord_sf(
    crs = st_crs(3338),
    xlim = x_limits,
    ylim = y_limits,
    expand = FALSE
  ) +
  # Add plot title
  ggtitle('a. Vegetation Types (1:2,500 scale)') +
  # Add plot styling
  guides(fill = guide_legend(byrow = TRUE)) +       
  theme_minimal() +
  theme(plot.margin = margin(2,2,2,2), 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_line(colour = 'gray'),
        plot.title = element_text(size = font_size)
  )

# Create vegetation type-complex map panel (top right)
vegcomplex_plot = ggplot() +
  # Add vegetation complex raster
  geom_spatraster(data = vegcomplex_crop) +
  scale_fill_manual(values = color_palette,               
                    breaks = type_labels,           
                    na.translate = FALSE, 
                    name = 'Vegetation Class') +
  # Add coastal zone overlay
  geom_sf(data = coast_data, color = 'black', fill = NA, linewidth = 0.8) +
  # Add installation boundary overlay
  geom_sf(data = installation_data, color = 'yellow', fill = NA, linewidth = 0.8) +
  # Adjust the plot to the horizontal and vertical installation bounds
  coord_sf(
    crs = st_crs(3338),
    xlim = x_limits,
    ylim = y_limits,
    expand = FALSE
  ) +
  # Add plot title
  ggtitle('b. Vegetation Types-Complexes (1:10,000 scale)') +
  # Add plot styling
  guides(fill = guide_legend(byrow = TRUE)) +       
  theme_minimal() +
  theme(plot.margin = margin(2,2,2,2), 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_line(colour = 'gray'),
        plot.title = element_text(size = font_size)
  )

#### CONSTRUCT MULTI-PANEL LEGEND
####____________________________________________________

# Remove legends from individual plots
vegtype_plot = vegtype_plot + theme(legend.position = 'none')
vegcomplex_plot = vegcomplex_plot + theme(legend.position = 'none')

# Alphabetize legend items
alphabetical_labels = sort(as.character(type_labels))

# Create a data frame of sorted legend items
legend_data = data.frame(
  class = factor(alphabetical_labels, levels = alphabetical_labels)
)

# Build the legend plot
legend_plot = ggplot() +
  geom_rect(data = legend_data, 
            aes(xmin = x_limits[1], xmax = x_limits[1], 
                ymin = y_limits[1], ymax = y_limits[1], 
                fill = class), 
            alpha = 0) +
  scale_fill_manual(values = color_palette,               
                    breaks = alphabetical_labels, # Assign legend order
                    drop = FALSE,                 # Ensures no items are dropped
                    name = NULL) +                # Suppress the normal legend title
  coord_sf(
    crs = st_crs(3338),
    xlim = x_limits,
    ylim = y_limits,
    expand = FALSE,
    clip = "off"
  ) +
  ggtitle('Vegetation Types and Complexes') + 
  guides(fill = guide_legend(ncol = 1, byrow = TRUE, title = NULL, override.aes = list(alpha = 1))) +
  theme_minimal() +
  theme(plot.margin = margin(2,2,2,2), 
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        plot.title = element_text(size = font_size),
        legend.position = c(0, 1),
        legend.justification = c(0, 1),
        legend.text = element_text(size = font_size - 2),
        legend.spacing.y = unit(2, 'pt'),
        legend.key.height = unit(1.1, 'cm'),
        legend.key.width = unit(1.1, 'cm'),
        legend.background = element_blank(),
        legend.box.background = element_blank()
  )

#### MERGE AND EXPORT PLOTS
####____________________________________________________

# Create the top row (vegetation maps)
row1 = plot_grid(vegtype_plot, vegcomplex_plot, ncol = 2, align = 'hv')

# Create middle row (imagery and legend)
row2 = plot_grid(imagery_plot, legend_plot, ncol = 2, align = 'hv')

# Combine the rows vertically and append a NULL row at the bottom 
combine_plot = plot_grid(row1, row2, NULL, 
                         ncol = 1, 
                         rel_heights = c(1, 1, 1))

# Export plot
ggsave(figure_output,
       plot = combine_plot,
       device = 'jpeg',
       path = NULL,
       scale = 2,
       width = 6.5,     
       height = 10,     
       units = 'in',
       dpi = 600,
       limitsize = TRUE)

File: 02_icy_cape/archive/11_installation_maps.py
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Icy Cape map results
# Author: Timm Nawrocki, Alaska Center for Conservation Science (Translated to Python)
# Last Updated: 2026-04-24
# Usage: Must be executed in a Python 3.9+ environment with geopandas, rioxarray, and matplotlib.
# Description: 'Icy Cape map results' creates a map figure for publication that shows the imagery, 0.5 m vegetation types, and 2 m vegetation types.
# ---------------------------------------------------------------------------

import os
import re
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.colors import ListedColormap, BoundaryNorm
import geopandas as gpd
import rioxarray

# Set global font size
font_size = 20

#### SET UP DIRECTORIES, FILES, AND FIELDS
####____________________________________________________

# Set root directory
drive = 'C:/'
root_folder = 'ACCS_Work'

# Define folder structure
project_folder = os.path.join(drive, root_folder, 'Projects/VegetationEcology/DoD_Navy_Arctic/Data')
imagery_folder = os.path.join(project_folder, 'Data_Input/imagery_data')
region_folder = os.path.join(project_folder, 'Data_Input/region_data')
vegetation_folder = os.path.join(project_folder, 'Data_Output/vegetation_data')
output_folder = os.path.join(project_folder, 'Data_Output/summary_results')

# Define input files
imagery_input = os.path.join(imagery_folder, 'IcyCape_Imagery_20200714_0.5m_3338.tif')
installation_input = os.path.join(region_folder, 'IcyCape_Installation_3338.shp')
coast_input = os.path.join(region_folder, 'IcyCape_CoastalZone_3338.shp')
vegtype_input = os.path.join(vegetation_folder, 'IcyCape_Vegetation_2020_0.5m_3338.tif')
vegcomplex_input = os.path.join(vegetation_folder, 'IcyCape_Vegetation_2020_900mmu_2m_3338.tif')

# Define output files
figure_output = os.path.join(output_folder, 'IcyCape_Map_Results_py.jpg')


#### DEFINE FUNCTIONS
####____________________________________________________

def get_color_palette(color_path):
    """
    Parses a .clr colormap file and returns a dictionary mapping
    values to hex colors and labels. (In Python, extracting the RAT directly
    from the TIFF requires reading the sidecar .vat.dbf, so we extract labels
    from the .clr file if present, or fallback to class values).
    """
    color_dict = {}
    with open(color_path, 'r') as f:
        for line in f:
            # Keep all lines that start with a digit
            if re.match(r'^\d', line):
                parts = line.strip().split()
                val = int(parts[0])
                r, g, b = int(parts[1]), int(parts[2]), int(parts[3])
                hex_color = f'#{r:02x}{g:02x}{b:02x}'

                # Assume label is at the end of the line if provided, else use the VALUE
                label = " ".join(parts[4:]) if len(parts) > 4 else f"Class {val}"
                color_dict[val] = {'hex': hex_color, 'label': label}

    return color_dict


def plot_categorical_raster(ax, raster_da, palette_dict, extent):
    """Helper function to plot a categorical SpatRaster equivalent in matplotlib."""
    raster_array = raster_da.squeeze().values

    # Get unique values from the raster
    unique_vals = np.unique(raster_array[~np.isnan(raster_array)])

    # Build matplotlib colormap and normalization
    colors = [palette_dict.get(v, {'hex': '#000000'})['hex'] for v in unique_vals]
    cmap = ListedColormap(colors)

    # BoundaryNorm requires boundaries between values
    bounds = list(unique_vals) + [unique_vals[-1] + 1]
    norm = BoundaryNorm(bounds, cmap.N)

    ax.imshow(raster_array, cmap=cmap, norm=norm, extent=extent, interpolation='none')


#### PREPARE VECTOR DATA & DEFINE MAP EXTENT
####____________________________________________________

# Read shapefiles and enforce EPSG:3338
coast_data = gpd.read_file(coast_input).to_crs(epsg=3338)
installation_data = gpd.read_file(installation_input).to_crs(epsg=3338)

# Buffer the installation polygon by 50 meters
installation_buffered = installation_data.buffer(50)

# Create a bounding box based on the 50 m buffered installation bounds
# total_bounds returns [xmin, ymin, xmax, ymax]
bounds = installation_buffered.total_bounds
x_limits = (bounds[0], bounds[2])
y_limits = (bounds[1], bounds[3])

# Matplotlib extent format: [xmin, xmax, ymin, ymax]
map_extent = [bounds[0], bounds[2], bounds[1], bounds[3]]

#### PREPARE RASTER DATA & COLORMAPS
####____________________________________________________

# Read and crop imagery (rioxarray uses minx, miny, maxx, maxy)
imagery_raster = rioxarray.open_rasterio(imagery_input, masked=True)
imagery_crop = imagery_raster.rio.clip_box(minx=bounds[0], miny=bounds[1], maxx=bounds[2], maxy=bounds[3])

# Read and crop vegetation type map
vegtype_raster = rioxarray.open_rasterio(vegtype_input, masked=True)
vegtype_crop = vegtype_raster.rio.clip_box(minx=bounds[0], miny=bounds[1], maxx=bounds[2], maxy=bounds[3])

# Read and crop vegetation type-complex map
vegcomplex_raster = rioxarray.open_rasterio(vegcomplex_input, masked=True)
vegcomplex_crop = vegcomplex_raster.rio.clip_box(minx=bounds[0], miny=bounds[1], maxx=bounds[2], maxy=bounds[3])

# Generate palettes from .clr files
type_palette = get_color_palette(str(vegtype_input) + '.clr')
complex_palette = get_color_palette(str(vegcomplex_input) + '.clr')

# Combine the two color palettes
color_palette = {**type_palette, **complex_palette}

# Extract the labels present in the cropped raster extents
type_present = np.unique(vegtype_crop.squeeze().values)
complex_present = np.unique(vegcomplex_crop.squeeze().values)

# Remove NaNs
type_present = type_present[~np.isnan(type_present)]
complex_present = complex_present[~np.isnan(complex_present)]

# Combine the unique labels from both cropped rasters
all_present_vals = sorted(np.unique(np.concatenate((type_present, complex_present))))

#### BUILD INDIVIDUAL MAP PANELS
####____________________________________________________

# Set up the 2x2 grid figure
fig, axs = plt.subplots(2, 2, figsize=(16, 16), dpi=600, gridspec_kw={'wspace': 0.05, 'hspace': 0.05})

# Map axs indices to standard names
ax_vegtype = axs[0, 0]
ax_vegcomplex = axs[0, 1]
ax_imagery = axs[1, 0]
ax_legend = axs[1, 1]


# Shared styling helper
def style_map_axes(ax, title):
    coast_data.plot(ax=ax, edgecolor='black', facecolor='none', linewidth=0.8)
    installation_data.plot(ax=ax, edgecolor='yellow', facecolor='none', linewidth=0.8)
    ax.set_xlim(x_limits)
    ax.set_ylim(y_limits)
    ax.set_title(title, fontsize=font_size, pad=10, loc='left')
    ax.grid(color='gray', linestyle='-', linewidth=0.5)
    ax.set_xticks([])
    ax.set_yticks([])
    # Hide standard matplotlib border spines to mimic theme_minimal()
    for spine in ax.spines.values():
        spine.set_visible(False)


# --- 1. Create base imagery plot (Bottom Left) ---
# Extract bands 4, 3, 2 for CIR (in xarray, bands are 1-indexed: 4, 3, 2)
# Ensure your imagery's band metadata maps 4=NIR, 3=Red, 2=Green.
cir_array = imagery_crop.sel(band=[4, 3, 2]).values
cir_array = np.moveaxis(cir_array, 0, -1)  # Move bands to last dimension for matplotlib (H, W, 3)

# Apply linear stretch (max_col_value = 220 equivalent)
cir_array = cir_array / 220.0
cir_array = np.clip(cir_array, 0, 1)

ax_imagery.imshow(cir_array, extent=map_extent)
style_map_axes(ax_imagery, 'c. High-resolution (0.5 m) CIR imagery')

# Add label with semi-transparent background
bbox_props = dict(boxstyle="square,pad=0.3", fc="white", ec="none", alpha=0.7)
ax_imagery.text(x_limits[0] + 20, y_limits[0] + 20,
                'Imagery © 2020 Maxar Technologies Inc.',
                fontsize=font_size * 0.7,  # Scaled down slightly to match R scaling
                color='black',
                ha='left', va='bottom',
                bbox=bbox_props)

# --- 2. Create vegetation type map panel (Top Left) ---
plot_categorical_raster(ax_vegtype, vegtype_crop, color_palette, map_extent)
style_map_axes(ax_vegtype, 'a. Vegetation Types (1:2,500 scale)')

# --- 3. Create vegetation type-complex map panel (Top Right) ---
plot_categorical_raster(ax_vegcomplex, vegcomplex_crop, color_palette, map_extent)
style_map_axes(ax_vegcomplex, 'b. Vegetation Types-Complexes (1:10,000 scale)')

#### MERGE AND EXPORT PLOTS
####____________________________________________________

# --- 4. Render Legend (Bottom Right) ---
ax_legend.axis('off')  # Hide the axes entirely for the legend panel

# Create custom handles for the unified legend
legend_handles = []
for val in all_present_vals:
    if val in color_palette:
        color_info = color_palette[val]
        patch = mpatches.Patch(color=color_info['hex'], label=color_info['label'])
        legend_handles.append(patch)

# Add the legend to the empty bottom-right subplot
# labelspacing = 1.0 approx equals 2pt vertical spacing depending on font sizes
leg = ax_legend.legend(handles=legend_handles,
                       title='Vegetation Class',
                       loc='center',
                       fontsize=font_size - 2,
                       title_fontsize=font_size,
                       frameon=False,
                       ncol=2,  # Adjust columns (byrow equivalent)
                       labelspacing=0.8,  # Vertical spacing between items
                       handleheight=2.0,  # Key height (1.1 cm equivalent)
                       handlelength=2.0)  # Key width  (1.1 cm equivalent)

# Adjust title padding
leg._legend_box.align = "left"

# Export plot
plt.savefig(figure_output, format='jpg', dpi=600, bbox_inches='tight', pad_inches=0.1)
plt.close()